File: objectiveai-sdk-rs/src/functions/http.rs
//! HTTP functions for function management.

use crate::{HttpClient, HttpError};

/// Lists all functions accessible to the authenticated user.
pub async fn list_functions(
    client: &HttpClient,
    params: super::request::ListFunctionsRequest,
) -> Result<super::response::ListFunctionResponse, HttpError> {
    client
        .send_unary(reqwest::Method::POST, "functions/list", Some(params))
        .await
}

/// Retrieves a function definition.
pub async fn get_function(
    client: &HttpClient,
    params: super::request::GetFunctionRequest,
) -> Result<super::response::GetFunctionResponse, HttpError> {
    client
        .send_unary(reqwest::Method::POST, "functions", Some(params))
        .await
}

/// Gets usage statistics for a specific function.
pub async fn get_function_usage(
    client: &HttpClient,
    params: super::request::GetFunctionRequest,
) -> Result<super::response::UsageFunctionResponse, HttpError> {
    client
        .send_unary(reqwest::Method::POST, "functions/usage", Some(params))
        .await
}

/// Lists all function-profile pairs accessible to the authenticated user.
pub async fn list_function_profile_pairs(
    client: &HttpClient,
    params: super::request::ListFunctionProfilePairsRequest,
) -> Result<super::response::ListFunctionProfilePairResponse, HttpError> {
    client
        .send_unary(reqwest::Method::POST, "functions/profiles/pairs/list", Some(params))
        .await
}

/// Gets usage statistics for a specific function-profile pair.
pub async fn get_function_profile_pair_usage(
    client: &HttpClient,
    params: super::request::GetFunctionProfilePairUsageRequest,
) -> Result<super::response::UsageFunctionProfilePairResponse, HttpError> {
    client
        .send_unary(reqwest::Method::POST, "functions/profiles/pairs/usage", Some(params))
        .await
}

File: objectiveai-sdk-py/objectiveai_sdk/functions/http.py
"""HTTP functions for function management endpoints."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from objectiveai_sdk.client import ObjectiveAI
    from objectiveai_sdk.functions import (
        GetFunctionProfilePairUsageRequest,
        GetFunctionRequest,
        GetFunctionResponse,
        ListFunctionProfilePairResponse,
        ListFunctionProfilePairsRequest,
        ListFunctionResponse,
        ListFunctionsRequest,
        UsageFunctionProfilePairResponse,
        UsageFunctionResponse,
    )


async def list_functions(
    client: ObjectiveAI, params: ListFunctionsRequest,
) -> ListFunctionResponse:
    """List all functions accessible to the authenticated user."""
    return await client.post_unary("functions/list", params)


async def get_function(
    client: ObjectiveAI, params: GetFunctionRequest,
) -> GetFunctionResponse:
    """Retrieve a function definition."""
    return await client.post_unary("functions", params)


async def get_function_usage(
    client: ObjectiveAI, params: GetFunctionRequest,
) -> UsageFunctionResponse:
    """Retrieve usage statistics for a specific function."""
    return await client.post_unary("functions/usage", params)


async def list_function_profile_pairs(
    client: ObjectiveAI, params: ListFunctionProfilePairsRequest,
) -> ListFunctionProfilePairResponse:
    """List all function-profile pairs accessible to the authenticated user."""
    return await client.post_unary("functions/profiles/pairs/list", params)


async def get_function_profile_pair_usage(
    client: ObjectiveAI, params: GetFunctionProfilePairUsageRequest,
) -> UsageFunctionProfilePairResponse:
    """Retrieve usage statistics for a specific function-profile pair."""
    return await client.post_unary("functions/profiles/pairs/usage", params)
